docs/remote.md

```diff
@@ -1,18 +1,191 @@
+
+## Protocol
+
+The remote driver protocol is a set of RPCs, issued as HTTP POSTs with JSON payloads. The proxy issues requests, and the remote driver process is expected to respond usually with a JSON payload of its own, although in some cases these are empty maps.
+
+### Errors
+
+If the remote process cannot decode, or otherwise detects a syntactic problem with the HTTP request or payload, it must respond with an HTTP error status (4xx or 5xx).
+
+If the remote process can decode the request, but cannot complete the operation, it must send a response in the form
+
+    {
+        "Err": string
+    }
+
+The string value supplied may appear in logs, so should not include confidential information.
+
+### Handshake
+
+When loaded, a remote driver process receives an HTTP POST on the URL `/Plugin.Activate` with no payload. It must respond with a manifest of the form
+
+    {
+        "Implements": ["NetworkDriver"]
+    }
+
+Other entries in the list value are allowed; `"NetworkDriver"` indicates that the plugin should be registered with LibNetwork as a driver.
+
+### Create network
+
+When the proxy is asked to create a network, the remote process shall receive a POST to the URL `/NetworkDriver.CreateNetwork` of the form
+
+    {
+        "NetworkID": string,
+        "Options": {
+            ...
+        }
+    }
+
+The `NetworkID` value is generated by LibNetwork. The `Options` value is the arbitrary map given to the proxy by LibNetwork.
+
+The response indicating success is empty:
+
+    `{}`
+
+### Delete network
+
+When a network owned by the remote driver is deleted, the remote process shall receive a POST to the URL `/NetworkDriver.DeleteNetwork` of the form
+
+    {
+        "NetworkID": string
+    }
+
+The success response is empty:
+
+    {}
+
+### Create endpoint
+
+When the proxy is asked to create an endpoint, the remote process shall receive a POST to the URL `/NetworkDriver.CreateEndpoint` of the form
+
+    {
+        "NetworkID": string,
+        "EndpointID": string,
+        "Options": {
+            ...
+        },
+        "Interfaces": [{
+            "ID": int,
+            "Address": string,
+            "AddressIPv6": string,
+            "MacAddress": string
+        }, ...]
+    }
+
+The `NetworkID` is the generated identifier for the network to which the endpoint belongs; the `EndpointID` is a generated identifier for the endpoint.
+
+`Options` is an arbitrary map as supplied to the proxy.
+
+The `Interfaces` value is a list with values of the form given. The fields in the `Interfaces` entries may be empty; and the `Interfaces` list itself may be empty. If supplied, `Address` is an IPv4 address and subnet in CIDR notation; e.g., `"192.168.34.12/16"`. If supplied, `AddressIPv6` is an IPv6 address and subnet in CIDR notation. `MacAddress` is a MAC address as a string; e.g., `"6e:75:32:60:44:c9"`.
```

Comment:
It would be helpful to explicitly describe under what circumstances the `Interfaces` list will be non-empty, and what the driver is expected to do (e.g. not just what form the information takes).

E.g. If the end user requests a specific IP address to be assigned to an endpoint, the Interface list will contain an interface with that IP address.  If the interface is provided, the driver should attempt to assign the requested IP addresses from it's IPAM subsystem.  If the requested IP address cannot be allocated, the driver should return an error.

Or whatever is actually the case!